components/eamxx/src/dynamics/homme/homme_grids_manager.cpp

```diff
@@ -268,6 +283,120 @@ build_pg_codes () {
   m_pg_codes["GLL"]["Twin"] = 10;
   m_pg_codes["PG2"]["None"] =  2;
   m_pg_codes["PG2"]["Twin"] = 12;
-} 
+}
+
+namespace {
+
+// Helper function for creating mirror device view used to set geometry data.
+const AbstractGrid::geo_view_type cmvc(const AbstractGrid::geo_view_h_type host_view) {
+  return Kokkos::create_mirror_view_and_copy(DefaultDevice(), host_view);
+}
+
+}
+
+void HommeGridsManager::initialize_vertical_coordinates(const nonconstgrid_ptr_type &dyn_grid) const {
+  using vos_t = std::vector<std::string>;
+  using namespace ShortFieldTagsNames;
+  using geo_view_host = AbstractGrid::geo_view_type::HostMirror;
+
+  const int nlev_int = HOMMEXX_NUM_INTERFACE_LEV;
+  const int nlev_mid = HOMMEXX_NUM_PHYSICAL_LEV;
+
+  // Read vcoords into host views
+  ekat::ParameterList vcoord_reader_pl;
+  vcoord_reader_pl.set("Filename",m_params.get<std::string>("vertical_coordinate_filename"));
+  vcoord_reader_pl.set<vos_t>("Field Names",{"hyai","hybi","hyam","hybm"});
+  std::map<std::string,geo_view_host> host_views = {
+    { "hyai", geo_view_host("hyai",nlev_int) },
+    { "hybi", geo_view_host("hybi",nlev_int) },
+    { "hyam", geo_view_host("hyam",nlev_mid) },
+    { "hybm", geo_view_host("hybm",nlev_mid) }
+  };
+  std::map<std::string,FieldLayout> layouts = {
+    { "hyai", FieldLayout({ILEV},{nlev_int}) },
+    { "hybi", FieldLayout({ILEV},{nlev_int}) },
+    { "hyam", FieldLayout({LEV}, {nlev_mid}) },
+    { "hybm", FieldLayout({LEV}, {nlev_mid}) }
+  };
+
+  AtmosphereInput vcoord_reader(m_comm,vcoord_reader_pl);
+  vcoord_reader.init(dyn_grid,host_views,layouts);
+  vcoord_reader.read_variables();
+  vcoord_reader.finalize();
+
+  // Pass host views data to hvcoord init function
+  const auto ps0 = Homme::PhysicalConstants::p0;
+  prim_set_hvcoords_f90 (ps0,
+                         host_views["hyai"].data(),
+                         host_views["hybi"].data(),
+                         host_views["hyam"].data(),
+                         host_views["hybm"].data());
+
+  // Set geometry data
+  dyn_grid->set_geometry_data("hyai", cmvc(host_views["hyai"]));
+  dyn_grid->set_geometry_data("hybi", cmvc(host_views["hybi"]));
+  dyn_grid->set_geometry_data("hyam", cmvc(host_views["hyam"]));
+  dyn_grid->set_geometry_data("hybm", cmvc(host_views["hybm"]));
+}
+
+void HommeGridsManager::load_topography (const nonconstgrid_ptr_type& phys_grid,
+                                         const ci_string&             type) const {
+  using geo_view_device = AbstractGrid::geo_view_type;
+  using geo_view_host   = geo_view_device::HostMirror;
+
+  const int nlcols = phys_grid->get_num_local_dofs();
+
+  if (type == "PG2") {
+
+    // For PG2, homme AD interface will compute
```

Comment:
I'm not sure I fully understand why we can't load it. For PG2 cases, isn't the topo file storing both`phis` and `phis_d`? Or are we computing `phis` remapping `phis_d`?

Reply:
We could accurately load it, but then `fv_phys` code in `HommeDynamics` would immediately overwrite it.

The procedure is 
1. Load phis on GLL
2. Remap phys_gll -> phys_dyn
3. Compute phys_pg2 in `fv_phys` code

Reply:
Maybe in the future we want to load phis on pg2? I can change it to load phis for both GLL and PG2, and then just allow it to be overwritten. It's a bit more computation, but only in the setup phase so no big deal.

Reply:
No, we should not load phis_phys. It's in that file for use in other contexts, such as analysis, that don't have access to the projection routine. Load just phis_gll, and the projection will occur, as it already does, during the rest of initialization.

Reply:
I should add, as a reminder, that in the future some physics parameterizations may need data that exist only on the physics grid in the topo file, e.g., the SGH fields. (These are not L2 projections from a dynamics-grid field.)

Reply:
I reverted back to not loading PG2 topo, but I've kept the distinction between pg2 and non-pg2 files for topography_filename.